File: examples/python/arkode/ark_adr1d_extsts.py
#!/usr/bin/env python3
# -----------------------------------------------------------------
# Programmer(s): Daniel R. Reynolds @ UMBC
# -----------------------------------------------------------------
# SUNDIALS Copyright Start
# Copyright (c) 2025-2026, Lawrence Livermore National Security,
# University of Maryland Baltimore County, and the SUNDIALS contributors.
# Copyright (c) 2013-2025, Lawrence Livermore National Security
# and Southern Methodist University.
# Copyright (c) 2002-2013, Lawrence Livermore National Security.
# All rights reserved.
#
# See the top-level LICENSE and NOTICE files for details.
#
# SPDX-License-Identifier: BSD-3-Clause
# SUNDIALS Copyright End
# -----------------------------------------------------------------
# Python version of the C++ example
#   examples/arkode/CXX_serial/ark_adr1d_extsts.cpp
#
# This example solves a one-dimensional advection-diffusion-reaction
# system with three chemical species,
#
#   u_t = -c u_x + d u_xx + A - (w + 1) * u + v * u^2
#   v_t = -c v_x + d v_xx + w * u - v * u^2
#   w_t = -c w_x + d w_xx + (B - w) / eps - w * u
#
# on x in [0, 1], with stationary boundary conditions. The time
# integration uses an ExtSTS method:
#
#   diffusion: explicit inner STS method
#   advection: explicit slow RHS
#   reaction : implicit slow RHS
#
# The spatial derivatives use second-order centered differences on a
# uniform mesh.  The example stores the state as a serial N_Vector, but
# the callbacks view that vector as an (nx, 3) NumPy array so that each
# row is one mesh point and the columns are the u, v, and w species.
# -----------------------------------------------------------------

import argparse
import sys
import tempfile
from pathlib import Path

import numpy as np
import sundials4py.arkode as ark
import sundials4py.core as sun


class ADR1DExtSTSProblem:
    """Advection-diffusion-reaction problem data and ARKODE callbacks."""

    def __init__(
        self, c=1.0e-2, d=1.0e-1, A=0.6, B=2.0, eps=1.0e-2, tf=3.0, xl=0.0, xu=1.0, nx=512
    ):
        self.c = c
        self.d = d
        self.A = A
        self.B = B
        self.eps = eps
        self.tf = tf
        self.xl = xl
        self.xu = xu
        self.nx = nx

        self.dx = (xu - xl) / (nx - 1)
        self.neq = 3 * nx

    def view2d(self, yvec):
        """Return a zero-copy (mesh point, species) view of an N_Vector."""
        return sun.N_VGetArrayPointer(yvec).reshape((self.nx, 3))

    def set_initial_condition(self, yvec):
        y = self.view2d(yvec)
        x = np.linspace(self.xl, self.xu, self.nx)
        perturbation = 0.1 * np.sin(np.pi * x)

        y[:, 0] = self.A + perturbation
        y[:, 1] = self.B / self.A + perturbation
        y[:, 2] = self.B + perturbation
        return 0

    def f_advection(self, t, yvec, fvec, user_data):
        y = self.view2d(yvec)
        f = self.view2d(fvec)

        # Centered first derivative for the advection term.  The boundary
        # rows are left at zero to impose stationary boundary conditions.
        f[0, :] = 0.0
        f[-1, :] = 0.0
        f[1:-1, :] = -self.c * (y[2:, :] - y[:-2, :]) / (2.0 * self.dx)
        return 0

    def f_diffusion(self, t, yvec, fvec, user_data):
        y = self.view2d(yvec)
        f = self.view2d(fvec)

        # Centered second derivative for diffusion.  This is the fast RHS
        # evolved by the inner STS method.
        f[0, :] = 0.0
        f[-1, :] = 0.0
        f[1:-1, :] = self.d * (y[:-2, :] - 2.0 * y[1:-1, :] + y[2:, :]) / (self.dx * self.dx)
        return 0

    def f_reaction(self, t, yvec, fvec, user_data):
        y = self.view2d(yvec)
        f = self.view2d(fvec)

        # Reaction terms are local to each mesh point and are treated
        # implicitly by the outer ExtSTS method.
        f[0, :] = 0.0
        f[-1, :] = 0.0
        u = y[1:-1, 0]
        v = y[1:-1, 1]
        w = y[1:-1, 2]

        f[1:-1, 0] = self.A - (w + 1.0) * u + v * u * u
        f[1:-1, 1] = w * u - v * u * u
        f[1:-1, 2] = (self.B - w) / self.eps - w * u
        return 0

    def diffusion_domeig(self, t, yvec, fnvec, user_data, temp1, temp2, temp3):
        # Dominant eigenvalue estimate for the finite-difference diffusion
        # operator. The STS method uses this to choose stable inner steps.
        lambdaR = -4.0 * self.d / (self.dx * self.dx)
        lambdaI = 0.0
        return 0, lambdaR, lambdaI


def parse_args(argv=None):
    parser = argparse.ArgumentParser(
        description="1D advection-diffusion-reaction problem using ARKODE ExtSTS"
    )

    parser.add_argument("--c", type=float, default=1.0e-2, help="advection speed")
    parser.add_argument("--d", type=float, default=1.0e-1, help="diffusion constant")
    parser.add_argument("--A", type=float, default=0.6, help="reaction parameter A")
    parser.add_argument("--B", type=float, default=2.0, help="reaction parameter B")
    parser.add_argument("--eps", type=float, default=1.0e-2, help="stiffness parameter")
    parser.add_argument("--tf", type=float, default=3.0, help="final time")
    parser.add_argument("--xl", type=float, default=0.0, help="domain lower boundary")
    parser.add_argument("--xu", type=float, default=1.0, help="domain upper boundary")
    parser.add_argument("--nx", type=int, default=512, help="number of mesh points")

    parser.add_argument(
        "--sts_method", choices=("RKC", "RKL"), default="RKC", help="inner STS method"
    )
    parser.add_argument(
        "--mri_method", default="ARKODE_IMEX_MRI_GARK_GKC21", help="MRI coupling table name"
    )
    parser.add_argument("--rtol", type=float, default=1.0e-4, help="relative tolerance")
    parser.add_argument("--atol", type=float, default=1.0e-9, help="absolute tolerance")
    parser.add_argument("--fixed_h", type=float, default=0.0, help="fixed outer step size")
    parser.add_argument(
        "--maxsteps", type=int, default=10000, help="maximum internal steps between output times"
    )
    parser.add_argument("--nout", type=int, default=10, help="number of output times")

    args = parser.parse_args(argv)

    if args.nx < 3:
        parser.error("--nx must be at least 3")
    if args.xu <= args.xl:
        parser.error("--xu must be greater than --xl")
    if args.tf <= 0.0:
        parser.error("--tf must be positive")
    if args.eps <= 0.0:
        parser.error("--eps must be positive")
    if args.rtol <= 0.0 or args.atol <= 0.0:
        parser.error("--rtol and --atol must be positive")
    if args.fixed_h < 0.0:
        parser.error("--fixed_h must be nonnegative")
    if args.maxsteps <= 0:
        parser.error("--maxsteps must be positive")
    if args.nout <= 0:
        parser.error("--nout must be positive")

    return args


def print_setup(problem, args):
    print("\nProblem parameters and options:")
    print(" --------------------------------- ")
    print(f"  c                = {problem.c}")
    print(f"  d                = {problem.d}")
    print(f"  A                = {problem.A}")
    print(f"  B                = {problem.B}")
    print(f"  eps              = {problem.eps}")
    print(" --------------------------------- ")
    print(f"  tf               = {problem.tf}")
    print(f"  xl               = {problem.xl}")
    print(f"  xu               = {problem.xu}")
    print(f"  nx               = {problem.nx}")
    print(f"  dx               = {problem.dx}")
    print(" --------------------------------- ")
    print("  integrator       = ExtSTS")
    print("  advection        = Explicit")
    print("  diffusion        = Explicit STS")
    print("  reaction         = Implicit")
    print(f"  rtol             = {args.rtol}")
    print(f"  atol             = {args.atol}")
    print(f"  fixed h          = {args.fixed_h}")
    print(" --------------------------------- ")
    print(f"  MRI method       = {args.mri_method}")
    print(f"  STS method       = {args.sts_method}")
    print(" --------------------------------- \n")


def rms_norm(yvec, neq):
    return np.sqrt(sun.N_VDotProd(yvec, yvec) / neq)


def print_solution_header():
    print("          t                    ||y||_rms")
    print(" -----------------------------------------------")


def print_solution(t, yvec, neq):
    print(f" {t:22.15e} {rms_norm(yvec, neq):22.15e}")


def print_solution_footer():
    print(" -----------------------------------------------\n")


def stats_text(arkode_mem):
    with tempfile.NamedTemporaryFile(
        prefix="ark_adr1d_extsts_", suffix=".txt", delete=False
    ) as tmp:
        stats_path = Path(tmp.name)

    try:
        status, file_ptr = sun.SUNFileOpen(str(stats_path), "w")
        assert status == sun.SUN_SUCCESS
        status = ark.ARKodePrintAllStats(arkode_mem.get(), file_ptr, sun.SUN_OUTPUTFORMAT_TABLE)
        assert status == ark.ARK_SUCCESS

        file_ptr = None
        return stats_path.read_text()
    finally:
        stats_path.unlink(missing_ok=True)


def print_stats(arkode, sts):
    print("\nExtSTS Integrator:")
    print(stats_text(arkode), end="")

    print("\nInner STS Method:")
    print(stats_text(sts), end="")


def load_mri_coupling(method):
    method_value = getattr(ark, method, method)

    if isinstance(method_value, int):
        return ark.MRIStepCoupling_LoadTable(method_value)

    return ark.MRIStepCoupling_LoadTableByName(method_value)


def main(argv=None):
    args = parse_args(argv)

    if not hasattr(ark, "MRIStepCreateExtSTS") or not hasattr(ark, "MRIStepGetSTS"):
        raise RuntimeError("This example requires sundials4py with ARKODE ExtSTS bindings enabled")

    status, sunctx = sun.SUNContext_Create(sun.SUN_COMM_NULL)
    assert status == sun.SUN_SUCCESS

    problem = ADR1DExtSTSProblem(
        c=args.c,
        d=args.d,
        A=args.A,
        B=args.B,
        eps=args.eps,
        tf=args.tf,
        xl=args.xl,
        xu=args.xu,
        nx=args.nx,
    )

    print_setup(problem, args)

    y = sun.N_VNew_Serial(problem.neq, sunctx)
    assert y is not None
    problem.set_initial_condition(y)

    # Create the ExtSTS stepper.  Diffusion is the inner STS RHS, while
    # advection and reaction are explicit and implicit slow RHS functions.
    arkode = ark.MRIStepCreateExtSTS(
        problem.f_diffusion, problem.f_advection, problem.f_reaction, 0.0, y, sunctx
    )
    assert arkode is not None

    # Retrieve and configure the inner STS stepper created by ExtSTS.
    status, sts = ark.MRIStepGetSTS(arkode.get())
    assert status == ark.ARK_SUCCESS
    assert sts is not None

    sts_method = ark.ARKODE_LSRK_RKC_2 if args.sts_method == "RKC" else ark.ARKODE_LSRK_RKL_2
    status = ark.LSRKStepSetSTSMethod(sts.get(), sts_method)
    assert status == ark.ARK_SUCCESS

    status = ark.LSRKStepSetDomEigFn(sts.get(), problem.diffusion_domeig)
    assert status == ark.ARK_SUCCESS

    status = ark.ARKodeSStolerances(arkode.get(), args.rtol, args.atol)
    assert status == ark.ARK_SUCCESS

    if args.fixed_h > 0.0:
        status = ark.ARKodeSetFixedStep(arkode.get(), args.fixed_h)
        assert status == ark.ARK_SUCCESS

    # The reaction RHS is implicit.  Providing a band matrix and band linear
    # solver is enough here; ARKODE will approximate the Jacobian internally.
    J = sun.SUNBandMatrix(problem.neq, 2, 2, sunctx)
    assert J is not None

    LS = sun.SUNLinSol_Band(y, J, sunctx)
    assert LS is not None

    status = ark.ARKodeSetLinearSolver(arkode.get(), LS, J)
    assert status == ark.ARK_SUCCESS

    coupling = load_mri_coupling(args.mri_method)
    assert coupling is not None

    status = ark.MRIStepSetCoupling(arkode.get(), coupling)
    assert status == ark.ARK_SUCCESS

    status = ark.ARKodeSetMaxNumSteps(arkode.get(), args.maxsteps)
    assert status == ark.ARK_SUCCESS

    t = 0.0
    dtout = problem.tf / args.nout
    tout = dtout

    print_solution_header()
    print_solution(t, y, problem.neq)

    for _ in range(args.nout):
        status, t = ark.ARKodeEvolve(arkode.get(), tout, y, ark.ARK_NORMAL)
        if status < 0:
            raise RuntimeError("ARKodeEvolve failed with status {status}")

        print_solution(t, y, problem.neq)
        tout = min(tout + dtout, problem.tf)

    print_solution_footer()

    print("Final integrator statistics:")
    print_stats(arkode, sts)


if __name__ == "__main__":
    main(sys.argv[1:])

File: examples/arkode/CXX_serial/ark_adr1d_extsts.cpp
/* -----------------------------------------------------------------------------
 * Programmer(s): Daniel R. Reynolds @ UMBC
 * -----------------------------------------------------------------------------
 * SUNDIALS Copyright Start
 * Copyright (c) 2025-2026, Lawrence Livermore National Security,
 * University of Maryland Baltimore County, and the SUNDIALS contributors.
 * Copyright (c) 2013-2025, Lawrence Livermore National Security
 * and Southern Methodist University.
 * Copyright (c) 2002-2013, Lawrence Livermore National Security.
 * All rights reserved.
 *
 * See the top-level LICENSE and NOTICE files for details.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * SUNDIALS Copyright End
 * -----------------------------------------------------------------------------
 * This example simulates the same problem as in
 * ark_advection_diffusion_reaction.cpp, by David J. Gardner @ LLNL,
 *
 *   u_t = -c u_x + d u_xx + A - (w + 1) * u + v * u^2
 *   v_t = -c v_x + d v_xx + w * u - v * u^2
 *   w_t = -c w_x + d w_xx + (B - w) / eps - w * u
 *
 * where u, v, and w represent the concentrations of chemical species, c = 0.01
 * is the advection speed, d = 0.1 is the diffusion rate, and the species with
 * constant concentration over time are A = 0.6 and B = 2.0.
 *
 * The problem is evolved for t in [0, 3] and x in [0, 1], with initial
 * conditions given by
 *
 *   u(0,x) =  A  + 0.1 * sin(pi * x)
 *   v(0,x) = B/A + 0.1 * sin(pi * x)
 *   w(0,x) =  B  + 0.1 * sin(pi * x)
 *
 * and stationary boundary conditions i.e.,
 *
 *   u_t(t,0) = u_t(t,1) = 0,
 *   v_t(t,0) = v_t(t,1) = 0,
 *   w_t(t,0) = w_t(t,1) = 0.
 *
 * The system is advanced in time with an Extended Super Time Stepping (ExtSTS)
 * method, wherein diffusion is treated explicitly using a STS method, advection
 * is treated explicitly using the ExtSTS method, and reaction is treated
 * implicitly using the ExtSTS method.
 *
 * Several command line options are available to change the problem parameters
 * and integrator settings. Use the flag --help for more information.
 * ---------------------------------------------------------------------------*/

#include "ark_adr1d_extsts.hpp"

int main(int argc, char* argv[])
{
  // SUNDIALS context object for this simulation
  sundials::Context ctx;

  // -----------------
  // Setup the problem
  // -----------------

  UserData udata;
  UserOptions uopts;

  vector<string> args(argv + 1, argv + argc);

  int flag = ReadInputs(args, udata, uopts, ctx);
  if (flag < 0)
  {
    cerr << "ERROR: ReadInputs returned " << flag << endl;
    return 1;
  }
  if (flag > 0) { return 0; }

  flag = PrintSetup(udata, uopts);
  if (check_flag(flag, "PrintSetup")) { return 1; }

  // Create state vector and set initial condition
  N_Vector y = N_VNew_Serial(udata.neq, ctx);
  if (check_ptr(y, "N_VNew_Serial")) { return 1; }

  flag = SetIC(y, udata);
  if (check_flag(flag, "SetIC")) { return 1; }

  // --------------------
  // Setup the integrator
  // --------------------

  // ARKODE memory structures
  void* arkode_mem = nullptr;
  void* sts_mem    = nullptr;

  // Matrix and linear solver for ExtSTS integrators
  SUNMatrix A        = nullptr;
  SUNLinearSolver LS = nullptr;

  // Create integrator
  ARKRhsFn fe_RHS;   // explicit RHS function
  ARKRhsFn fi_RHS;   // implicit RHS function
  ARKLsJacFn Ji_RHS; // implicit RHS Jacobian function

  fe_RHS = (udata.advection) ? f_advection : nullptr;
  fi_RHS = (udata.reaction) ? f_reaction : nullptr;
  Ji_RHS = (udata.reaction) ? J_reaction : nullptr;

  // -------------------------------
  // Setup the ExtSTS integrator
  // -------------------------------

  // Create MRIStep ExtSTS solver memory
  arkode_mem = MRIStepCreateExtSTS(f_diffusion, fe_RHS, fi_RHS, ZERO, y, ctx);
  if (check_ptr(arkode_mem, "MRIStepCreateExtSTS")) { return 1; }

  // Access inner LSRKStep solver
  flag = MRIStepGetSTS(arkode_mem, &sts_mem);
  if (check_flag(flag, "MRIStepGetSTS")) { return 1; }

  // Attach user data (this attaches to both MRIStep and LSRKStep)
  flag = ARKodeSetUserData(arkode_mem, &udata);
  if (check_flag(flag, "ARKodeSetUserData")) { return 1; }

  // Select STS method
  ARKODE_LSRKMethodType ststype = (uopts.sts_method == 0) ? ARKODE_LSRK_RKC_2
                                                          : ARKODE_LSRK_RKL_2;
  flag                          = LSRKStepSetSTSMethod(sts_mem, ststype);
  if (check_flag(flag, "LSRKStepSetSTSMethod")) { return 1; }

  // Set dominant eigenvalue function and frequency
  flag = LSRKStepSetDomEigFn(sts_mem, diffusion_domeig);
  if (check_flag(flag, "LSRKStepSetDomEigFn")) { return 1; }
  flag = LSRKStepSetDomEigFrequency(sts_mem, uopts.ls_setup_freq);
  if (check_flag(flag, "LSRKStepSetDomEigFrequency")) { return 1; }

  // Set fixed step size
  if (uopts.fixed_h > ZERO)
  {
    flag = ARKodeSetFixedStep(arkode_mem, uopts.fixed_h);
    if (check_flag(flag, "ARKodeSetFixedStep")) { return 1; }
  }

  // Specify tolerances
  flag = ARKodeSStolerances(arkode_mem, uopts.rtol, uopts.atol);
  if (check_flag(flag, "ARKodeSStolerances")) { return 1; }

  // If implicit, setup solvers
  if (udata.reaction)
  {
    // Create banded matrix
    A = SUNBandMatrix(udata.neq, 2, 2, ctx);
    if (check_ptr(A, "SUNBandMatrix")) { return 1; }

    // Create linear solver
    LS = SUNLinSol_Band(y, A, ctx);
    if (check_ptr(LS, "SUNLinSol_Band")) { return 1; }

    // Attach linear solver
    flag = ARKodeSetLinearSolver(arkode_mem, LS, A);
    if (check_flag(flag, "ARKodeSetLinearSolver")) { return 1; }

    // Attach Jacobian function
    flag = ARKodeSetJacFn(arkode_mem, Ji_RHS);
    if (check_flag(flag, "ARKodeSetJacFn")) { return 1; }

    // Set linear solver setup frequency
    flag = ARKodeSetLSetupFrequency(arkode_mem, uopts.ls_setup_freq);
    if (check_flag(flag, "ARKodeSetLSetupFrequency")) { return 1; }

    // Tighten implicit solver tolerances
    flag = ARKodeSetNonlinConvCoef(arkode_mem, 1.e-1);
    if (check_flag(flag, "ARKodeSetNonlinConvCoef")) { return 1; }
    flag = ARKodeSetEpsLin(arkode_mem, 1.e-1);
    if (check_flag(flag, "ARKodeSetEpsLin")) { return 1; }

    // Use "deduce implicit RHS" option
    flag = ARKodeSetDeduceImplicitRhs(arkode_mem, SUNTRUE);
    if (check_flag(flag, "ARKodeSetDeduceImplicitRhs")) { return 1; }
  }

  // Select ExtSTS method via MRIStepCoupling structure
  MRIStepCoupling C;
  if (uopts.build_mri_table)
  {
    ARKodeButcherTable B = nullptr;
    if (udata.reaction)
    {
      B = ARKodeButcherTable_LoadDIRKByName(uopts.mri_method.c_str());
    }
    else { B = ARKodeButcherTable_LoadERKByName(uopts.mri_method.c_str()); }
    if (check_ptr(B, "ARKodeButcherTable_Load")) { return 1; }
    C = MRIStepCoupling_MIStoMRI(B, B->q, B->p);
    if (check_ptr(C, "MRIStepCoupling_MIStoMRI")) { return 1; }
    ARKodeButcherTable_Free(B);
  }
  else
  {
    C = MRIStepCoupling_LoadTableByName(uopts.mri_method.c_str());
    if (check_ptr(C, "MRIStepCoupling_LoadTableByName")) { return 1; }
  }
  flag = MRIStepSetCoupling(arkode_mem, C);
  if (check_flag(flag, "MRIStepSetCoupling")) { return 1; }
  MRIStepCoupling_Free(C);

  // Set max steps between outputs
  flag = ARKodeSetMaxNumSteps(arkode_mem, uopts.maxsteps);
  if (check_flag(flag, "ARKodeSetMaxNumSteps")) { return 1; }

  // Tighten safety factor for time step selection
  flag = ARKodeSetSafetyFactor(arkode_mem, 0.8);
  if (check_flag(flag, "ARKodeSetSafetyFactor")) { return 1; }

  // Set stopping time
  flag = ARKodeSetStopTime(arkode_mem, udata.tf);
  if (check_flag(flag, "ARKodeSetStopTime")) { return 1; }

  // ----------------------
  // Evolve problem in time
  // ----------------------

  // Initial time, time between outputs, output time
  sunrealtype t     = ZERO;
  sunrealtype dTout = udata.tf / uopts.nout;
  sunrealtype tout  = dTout;

  // Initial output
  flag = OpenOutput(udata, uopts);
  if (check_flag(flag, "OpenOutput")) { return 1; }

  flag = WriteOutput(t, y, udata, uopts);
  if (check_flag(flag, "WriteOutput")) { return 1; }

  // Loop over output times
  for (int iout = 0; iout < uopts.nout; iout++)
  {
    // Evolve
    flag = ARKodeEvolve(arkode_mem, tout, y, &t, ARK_NORMAL);
    if (check_flag(flag, "ARKodeEvolve")) { return 1; }

    // Output solution
    flag = WriteOutput(t, y, udata, uopts);
    if (check_flag(flag, "WriteOutput")) { return 1; }

    // Update output time
    tout += dTout;
    tout = (tout > udata.tf) ? udata.tf : tout;
  }

  // Close output
  flag = CloseOutput(uopts);
  if (check_flag(flag, "CloseOutput")) { return 1; }

  // ------------
  // Output stats
  // ------------

  if (uopts.output)
  {
    cout << "Final integrator statistics:" << endl;
    cout << fixed << setprecision(6);
    cout << endl << "ExtSTS Integrator:" << endl;
    flag = ARKodePrintAllStats(arkode_mem, stdout, SUN_OUTPUTFORMAT_TABLE);
    if (check_flag(flag, "ARKodePrintAllStats")) { return -1; }
    cout << endl << "Inner STS Method:" << endl;
    flag = ARKodePrintAllStats(sts_mem, stdout, SUN_OUTPUTFORMAT_TABLE);
    if (check_flag(flag, "ARKodePrintAllStats")) { return -1; }
  }

  // --------
  // Clean up
  // --------

  N_VDestroy(y);
  SUNMatDestroy(A);
  SUNLinSolFree(LS);
  ARKodeFree(&arkode_mem);

  return 0;
}

// -----------------------------------------------------------------------------
// Functions called by the integrator
// -----------------------------------------------------------------------------

// Advection RHS function
int f_advection(sunrealtype t, N_Vector y, N_Vector f, void* user_data)
{
  // Access problem data
  UserData* udata = (UserData*)user_data;

  // Access data arrays
  sunrealtype* ydata = N_VGetArrayPointer(y);
  if (check_ptr(ydata, "N_VGetArrayPointer")) { return -1; }

  sunrealtype* fdata = N_VGetArrayPointer(f);
  if (check_ptr(fdata, "N_VGetArrayPointer")) { return -1; }

  // Compute advection RHS
  sunrealtype ul, ur;
  sunrealtype vl, vr;
  sunrealtype wl, wr;

  sunrealtype c = -ONE * udata->c / (TWO * udata->dx);

  N_VConst(ZERO, f);
  for (sunindextype i = 1; i < udata->nx - 1; i++)
  {
    ul = ydata[UIDX(i - 1)];
    ur = ydata[UIDX(i + 1)];

    vl = ydata[VIDX(i - 1)];
    vr = ydata[VIDX(i + 1)];

    wl = ydata[WIDX(i - 1)];
    wr = ydata[WIDX(i + 1)];

    fdata[UIDX(i)] = c * (ur - ul);
    fdata[VIDX(i)] = c * (vr - vl);
    fdata[WIDX(i)] = c * (wr - wl);
  }

  return 0;
}

// Diffusion RHS function
int f_diffusion(sunrealtype t, N_Vector y, N_Vector f, void* user_data)
{
  // Access problem data
  UserData* udata = (UserData*)user_data;

  // Access data arrays
  sunrealtype* ydata = N_VGetArrayPointer(y);
  if (check_ptr(ydata, "N_VGetArrayPointer")) { return -1; }

  sunrealtype* fdata = N_VGetArrayPointer(f);
  if (check_ptr(fdata, "N_VGetArrayPointer")) { return -1; }

  // Compute diffusion RHS
  sunrealtype ul, uc, ur;
  sunrealtype vl, vc, vr;
  sunrealtype wl, wc, wr;

  sunrealtype d = udata->d / (udata->dx * udata->dx);

  N_VConst(ZERO, f);
  for (sunindextype i = 1; i < udata->nx - 1; i++)
  {
    ul = ydata[UIDX(i - 1)];
    uc = ydata[UIDX(i)];
    ur = ydata[UIDX(i + 1)];

    vl = ydata[VIDX(i - 1)];
    vc = ydata[VIDX(i)];
    vr = ydata[VIDX(i + 1)];

    wl = ydata[WIDX(i - 1)];
    wc = ydata[WIDX(i)];
    wr = ydata[WIDX(i + 1)];

    fdata[UIDX(i)] = d * (ul - TWO * uc + ur);
    fdata[VIDX(i)] = d * (vl - TWO * vc + vr);
    fdata[WIDX(i)] = d * (wl - TWO * wc + wr);
  }

  return 0;
}

// Reaction RHS function
int f_reaction(sunrealtype t, N_Vector y, N_Vector f, void* user_data)
{
  // Access problem data
  UserData* udata = (UserData*)user_data;

  // Access data arrays
  sunrealtype* ydata = N_VGetArrayPointer(y);
  if (check_ptr(ydata, "N_VGetArrayPointer")) { return -1; }

  sunrealtype* fdata = N_VGetArrayPointer(f);
  if (check_ptr(fdata, "N_VGetArrayPointer")) { return -1; }

  // Compute reaction RHS
  sunrealtype u, v, w;

  N_VConst(ZERO, f);
  for (sunindextype i = 1; i < udata->nx - 1; i++)
  {
    u = ydata[UIDX(i)];
    v = ydata[VIDX(i)];
    w = ydata[WIDX(i)];

    fdata[UIDX(i)] = udata->A - (w + ONE) * u + v * u * u;
    fdata[VIDX(i)] = w * u - v * u * u;
    fdata[WIDX(i)] = ((udata->B - w) / udata->eps) - w * u;
  }

  return 0;
}

// Reaction Jacobian function
int J_reaction(sunrealtype t, N_Vector y, N_Vector fy, SUNMatrix J,
               void* user_data, N_Vector tmp1, N_Vector tmp2, N_Vector tmp3)
{
  // Access problem data
  UserData* udata = (UserData*)user_data;

  // Access data array
  sunrealtype* ydata = N_VGetArrayPointer(y);
  if (check_ptr(ydata, "N_VGetArrayPointer")) { return 1; }

  sunrealtype u, v, w;

  SUNMatZero(J);
  for (sunindextype i = 1; i < udata->nx - 1; i++)
  {
    u = ydata[UIDX(i)];
    v = ydata[VIDX(i)];
    w = ydata[WIDX(i)];

    // all vars wrt u
    SM_ELEMENT_B(J, UIDX(i), UIDX(i)) = -(w + ONE) + TWO * u * v;
    SM_ELEMENT_B(J, VIDX(i), UIDX(i)) = w - TWO * u * v;
    SM_ELEMENT_B(J, WIDX(i), UIDX(i)) = -w;

    // all vars wrt v
    SM_ELEMENT_B(J, UIDX(i), VIDX(i)) = u * u;
    SM_ELEMENT_B(J, VIDX(i), VIDX(i)) = -u * u;

    // all vars wrt w
    SM_ELEMENT_B(J, UIDX(i), WIDX(i)) = -u;
    SM_ELEMENT_B(J, VIDX(i), WIDX(i)) = u;
    SM_ELEMENT_B(J, WIDX(i), WIDX(i)) = (-ONE / udata->eps) - u;
  }

  return 0;
}

// Dominant eigenvalue function (for diffusion operator in LSRKStep)
int diffusion_domeig(sunrealtype t, N_Vector y, N_Vector fn,
                     sunrealtype* lambdaR, sunrealtype* lambdaI, void* user_data,
                     N_Vector temp1, N_Vector temp2, N_Vector temp3)
{
  // Access problem data
  UserData* udata = (UserData*)user_data;

  // Fill in spectral radius value
  *lambdaR = -SUN_RCONST(4.0) * udata->d / udata->dx / udata->dx;
  *lambdaI = SUN_RCONST(0.0);

  return 0;
}

// Compute the initial condition
int SetIC(N_Vector y, UserData& udata)
{
  sunrealtype* ydata = N_VGetArrayPointer(y);
  if (check_ptr(ydata, "N_VGetArrayPointer")) { return -1; }

  sunrealtype x, p;

  for (sunindextype i = 0; i < udata.nx; i++)
  {
    x              = udata.xl + i * udata.dx;
    p              = SUN_RCONST(0.1) * sin(PI * x);
    ydata[UIDX(i)] = udata.A + p;
    ydata[VIDX(i)] = udata.B / udata.A + p;
    ydata[WIDX(i)] = udata.B + p;
  }

  return 0;
}

//---- end of file ----
